cmd/thanos/main.go

```diff
@@ -212,6 +212,38 @@ func registerMetrics(mux *http.ServeMux, g prometheus.Gatherer) {
 	mux.Handle("/metrics", promhttp.HandlerFor(g, promhttp.HandlerOpts{}))
 }
 
+func registerHealthy(mux *http.ServeMux, logger log.Logger, component string) {
+	mux.HandleFunc("/-/healthy", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		if _, err := fmt.Fprintf(w, "Thanos %v is Healthy.\n", component); err != nil {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			level.Error(logger).Log("msg", "Could not write health check response.")
+		}
+	})
+}
+
+func registerReady(mux *http.ServeMux, logger log.Logger, component string, readinessFunc func() (bool, error)) {
+	mux.HandleFunc("/-/ready", func(w http.ResponseWriter, r *http.Request) {
+		ready, err := readinessFunc()
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to get readiness status.")
+			ready = false
+		}
+
+		if ready {
+			w.WriteHeader(http.StatusOK)
+			if _, err := fmt.Fprintf(w, "Thanos %v is ready.\n", component); err != nil {
+				level.Error(logger).Log("msg", "Could not write readiness check response.")
+			}
+			return
+		}
+		w.WriteHeader(http.StatusServiceUnavailable)
```

Comment:
Internal error maybe? Service Unavailable does not mean the pod is not unavilable, not ready. It is in context of this HTTP endpoint not available, and it encourage client to do backoff retry. I think 500 (internalError) might be more suitable here.